src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -588,77 +588,115 @@ ucs_status_t uct_dc_mlx5_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov,
-                                  uct_completion_t *comp)
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_dc_mlx5_ep_flush_cancel(uct_dc_mlx5_ep_t *ep, uint16_t *sn_p)
 {
-    uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_mlx5_iface_t);
-    uct_dc_mlx5_ep_t    *ep    = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
-    uct_ib_mlx5_md_t    *md    = ucs_derived_of(iface->super.super.super.super.md,
-                                                uct_ib_mlx5_md_t);
-    uint8_t pool_index         = uct_dc_mlx5_ep_pool_index(ep);
-    ucs_status_t        status;
-    uint16_t            sn;
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                                uct_dc_mlx5_iface_t);
+    uct_ib_mlx5_md_t *md       =
+            ucs_derived_of(iface->super.super.super.super.md,
+                           uct_ib_mlx5_md_t);
+    ucs_status_t status;
     UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
 
-    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
-        if (uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index)) {
-            UCT_TL_EP_STAT_FLUSH(&ep->super); /* no sends */
-            return UCS_OK;
-        }
-
-        return UCS_ERR_NO_RESOURCE; /* waiting for dci */
+    if (uct_dc_mlx5_iface_is_dci_rand(iface)) {
+        return UCS_ERR_UNSUPPORTED;
     }
 
-    if (!uct_dc_mlx5_iface_has_tx_resources(iface)) {
-        return UCS_ERR_NO_RESOURCE;
+    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
+        UCT_TL_EP_STAT_FLUSH(&ep->super);
+        return UCS_OK; /* nothing to cancel */
     }
 
-    if (!uct_dc_mlx5_iface_dci_ep_can_send(ep)) {
+    if (!uct_dc_mlx5_iface_has_tx_resources(iface) ||
+        !uct_dc_mlx5_iface_dci_has_tx_resources(iface, ep->dci)) {
         return UCS_ERR_NO_RESOURCE; /* cannot send */
     }
 
+    ucs_assert(ep->dci != UCT_DC_MLX5_EP_NO_DCI);
+
     status = uct_dc_mlx5_iface_flush_dci(iface, ep->dci);
     if (status == UCS_OK) {
         UCT_TL_EP_STAT_FLUSH(&ep->super);
         return UCS_OK; /* all sends completed */
     }
 
     ucs_assert(status == UCS_INPROGRESS);
-    ucs_assert(ep->dci != UCT_DC_MLX5_EP_NO_DCI);
 
     UCT_DC_MLX5_IFACE_TXQP_GET(iface, ep, txqp, txwq);
-    sn = txwq->sig_pi;
+    *sn_p = txwq->sig_pi;
 
-    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
-        UCT_DC_MLX5_CHECK_RES(iface, ep);
+    if (ep->flags & UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL) {
+        return UCS_INPROGRESS;
+    }
 
-        if (uct_dc_mlx5_iface_is_dci_rand(iface)) {
-            return UCS_ERR_UNSUPPORTED;
-        }
+    status = uct_ib_mlx5_modify_qp_state(md, &txwq->super, IBV_QPS_ERR);
+    if (status != UCS_OK) {
+        return status;
+    }
 
-        if (ep->flags & UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL) {
-            goto out;
-        }
+    uct_ib_mlx5_txwq_update_flags(txwq, UCT_IB_MLX5_TXWQ_FLAG_FAILED, 0);
+    ep->flags |= UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL;
+    *sn_p      = txwq->sw_pi;
```

Comment:
why need to reinit?

Reply:
it is initialized by sw_pi instead of sig_pi

Reply:
it seems that we could always wait for the last signaled WQE, i.e. `sig_pi` which is advanced after NO-OP is posted.
since there is no difference between:
- take sw_pi before calling signaled post_send
- take sig_pi after calling signaled post_send
their values should be equal, if I'm not mistaken

so, it should simplify the code much

Reply:
yes, sig_pi should be same as sw_pi before calling post(NOP)